src/t_cose_recipient_enc_esdh.c

```diff
@@ -0,0 +1,300 @@
+    struct t_cose_key       kek_handle;
+    size_t                  target_kek_len;
+    struct t_cose_recipient_enc_esdh *context;
+
+    context = (struct t_cose_recipient_enc_esdh *)me_x;
+
+    switch (context->esdh_suite.kw_id)
+    {
+    case T_COSE_ALGORITHM_A128KW:
+        target_kek_len = 16;
+        break;
+    case T_COSE_ALGORITHM_A192KW:
+        target_kek_len = 24;
+        break;
+    case T_COSE_ALGORITHM_A256KW:
+        target_kek_len = 32;
+        break;
+    default:
+        return(T_COSE_ERR_UNSUPPORTED_ENCRYPTION_ALG);
+    }
+
+    /* Create ephemeral key */
+    return_value = t_cose_crypto_generate_key(&ephemeral_key,
+                                              context->esdh_suite.curve_id);
+    if (return_value != T_COSE_SUCCESS) {
+        return(return_value);
+    }
+
+    /* Export pkR */
+    return_value = t_cose_crypto_export_public_key(
+                         context->pkR,
+                         (struct q_useful_buf) {.ptr=pkR, .len=pkR_len},
+                         &pkR_len);
+
+    if (return_value != T_COSE_SUCCESS) {
+        return(return_value);
+    }
+
+    /* Export pkE */
+    return_value = t_cose_crypto_export_public_key(
+                         ephemeral_key,
+                         (struct q_useful_buf) {.ptr=pkE, .len=pkE_len},
+                         &pkE_len);
+
+    if (return_value != T_COSE_SUCCESS) {
+        return(return_value);
+    }
+
+    /* Create COSE_recipient array */
+    QCBOREncode_OpenArray(cbor_encoder);
+
+    /* --- Make Info structure ---- */
+
+    /* Create structure for protected header */
+    QCBOREncode_Init(&protected_hdr_ctx, protected_hdr_buffer);
+    QCBOREncode_BstrWrap(&protected_hdr_ctx);
+    QCBOREncode_OpenMap(&protected_hdr_ctx);
+
+    QCBOREncode_AddInt64ToMapN(&protected_hdr_ctx,
+                               T_COSE_HEADER_PARAM_ALG,
+                               context->esdh_suite.ckd_id);
+
+    QCBOREncode_CloseMap(&protected_hdr_ctx);
+
+    QCBOREncode_CloseBstrWrap2(&protected_hdr_ctx,
+                               false,
+                               &protected_hdr);
+
+    ret = QCBOREncode_Finish(&protected_hdr_ctx, &protected_hdrC);
+
+    if (ret != QCBOR_SUCCESS) {
+        return(T_COSE_ERR_CBOR_NOT_WELL_FORMED);
+    }
+
+    return_value = create_info_structure(context->info->enc_alg,
+                                         context->info->sender_identity_type_id,
+                                         context->info->sender_identity,
+                                         context->info->recipient_identity_type_id,
+                                         context->info->recipient_identity,
+                                         protected_hdr,
+                                         context->info->enc_ctx->hash_cose_algorithm_id,
+                                         (struct q_useful_buf_c)
+                                         {.ptr = context->info->enc_ctx->extern_hash_buffer.ptr,
```

Comment:
Code like this is a signing that q_useful_buf is not being used well. There are several other places like this too.